src/conn/conn_dhandle.c

```diff
@@ -313,6 +313,16 @@ __wt_conn_dhandle_close(WT_SESSION_IMPL *session, bool final, bool mark_dead)
     btree = is_btree ? dhandle->handle : NULL;
 
     if (is_btree) {
+        /*
+         * !! Check if we have exclusive access here, else a transaction can start on this btree
+         * after this check !!
+         *
+         * Don't close the btree if there are changes not yet globally visible.
+         */
+        if (!WT_DHANDLE_IS_CHECKPOINT(dhandle) && (session != conn->sweep_session) &&
+          !__wt_txn_visible_all(session, btree->max_upd_txn, WT_TS_NONE))
```

Comment:
@tod-johnson-mongodb  - I think my review and your most recent changes might have crossed in flight. Sorry about that.

I think with the suggested change above, the plumbing of "check_visibility" should not be required any more, since the above tighter check will succeed in all cases, not just the drop case. I'm happy to hear if you think it's better to be explicit regardless - I could be convinced that is the case.